### PR TITLE
Tooling: Add volta versions

### PR DESCRIPTION
**Pull request by sirreal**
#### Changes proposed in this Pull Request

https://volta.sh/

This prepares the environment for use with Volta, a potential replacement for `nvm` (but more).

It sets a `volta` property in the root package.json. It also points _all the other package.json_ in the repo to extend that root config. This is under the assumption that all tooling should use the version in the root.

It was a little annoying to add a volta property to _every_ package.json, but that should be a one-time thing and doesn't seem like it should have any negative impact. That seems to be the way to do things in Volta at the moment. See https://github.com/volta-cli/volta/issues/906 and the [Workspaces docs](https://docs.volta.sh/advanced/workspaces).

#### Testing instructions

You'll need to set your system up for Volta. I have no idea if that will play well with nvm. I've removed nvm from my system completely.

From an arbitrary location like `~`, install some volta tools (that don't correspond to the versions in Calypso). For example:

- `volta install node@latest`
- `volta install typescript@beta`

Check the versions with `volta list`:

```none
$ volta list
⚡️ Currently active tools:

    Node: v15.6.0 (default)
    npm: v6.14.11 (default)
    Yarn: v1.22.10 (default)
    Tool binaries available:
        wp-env (default)
        neovim-node-host (default)
        tsc, tsserver (default)
        yarn-deduplicate (default)

$ node --version
v15.6.0
$ tsc --version
Version 4.2.0-dev
```

move to Calypso and check the versions:

```
$ volta list
⚡️ Currently active tools:

    Node: v12.20.1 (current @ /Users/jonsurrell/a8c/calypso/package.json)
    Yarn: v1.22.10 (current @ /Users/jonsurrell/a8c/calypso/package.json)
    Tool binaries available:
        wp-env (default)
        neovim-node-host (default)
        tsc, tsserver (current @ /Users/jonsurrell/a8c/calypso/package.json)
        yarn-deduplicate (default)

# Nice, volta picked up different versions! Let's check…

$ node --version
# installs on demand if necessary!!
v12.20.1

# It detects the package's version and prefers it!
$ tsc --version
Version 4.1.3
```

Repeat from another directory with a package.json such as `packages/*/` or `apps/editing-toolkit` or `client`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -315,5 +315,9 @@
 		"bunyan": "^1.8.14",
 		"chokidar": "^2.1.8",
 		"eslint-nibble": "^6.0.0"
+	},
+	"volta": {
+		"node": "12.19.1",
+		"yarn": "1.22.10"
```
Comment (sirreal) [inline package.json]: Out of date https://github.com/Automattic/wp-calypso/pull/48852

**Comment by sirreal** [comment]
Looks like renovate should support: https://github.com/renovatebot/renovate/issues/4262

**Review by noahtallen — APPROVED** [review]
I did a bit of testing with `nvm` + `volta`, and it looks like if NVM is configured in .the zshrc file, then it overrides volta. `nvm use` sets to 12.x, and that persists when changing to other directories. (e.g. the volta version is not used)

Screen Shot 2021-02-17 at 2 54 04 PM

When `nvm` is _not_ set in .zshrc, it works as expected:

Screen Shot 2021-02-17 at 2 56 40 PM

It appears that `nvm` still works fine with volta config available. So in other words, this shouldn't break anything for existing users :)

I'm all for this, as I've been getting frustrated by NVM. For example, if I try to do a git commit or push via Visual Studio Code's GUI, the precommit hooks will utilize the "default" node version rather than the local `nvm` node version, which sucks when different projects use different node versions.

After trying out the precommit hooks with volta, it looks like that problem is finally solved! My editor is using the node version for the project without any intervention. (Tried with both wp-calypso and gutenberg using your branches.)

File: package.json [hunk 2]
```diff
@@ -315,5 +315,9 @@
 		"bunyan": "^1.8.14",
 		"chokidar": "^2.1.8",
 		"eslint-nibble": "^6.0.0"
+	},
+	"volta": {
+		"node": "12.20.1",
+		"yarn": "1.22.10"
```
Comment (scinos) [inline package.json]: How does this interact with the version set in `.yarnrc`?

**Comment by scinos** [comment]
It is great that we are looking for alternatives to NVM if NVM is causing problems. Could you expand on the problems we are trying to solve with `volta`?

In any case, I think that we should only have one tool to do one thing. In other words, if we go ahead with this change, it should be a migration from `nvm` to `volta` with the aim of removing `nvm` from our system. Otherwise we are increasing the maintenance burden and the risk of `nvm

**Comment by lynnbgriffith** [comment]
> It is great that we are looking for alternatives to NVM if NVM is causing problems. Could you expand on the problems we are trying to solve with `volta`?

I found @noahtallen’s [comment on the related Gutenberg PR](https://github.com/WordPress/gutenberg/pull/28967#pullrequestreview-592707444) helpful in this regard. With volta, you no longer need to `nvm i && nvm use`. Instead, you should automatically use the correct node version for the project. That’s nice when on an interactive shell but perhaps more useful in the context of other tools that might otherwise end up using the incorrect version.

Dropping `nvm` in Calypso would be fine with me.

**Comment by noahtallen** [comment]
> Dropping nvm in Calypso would be fine with me.

What if we moved forward with this PR in the interim to test it further? That we way don't break existing `nvm` workflows, and we could try volta out.

**Comment by noahtallen** [comment]
One thing I noticed when testing this is that `yarn dev --sync` with ETK doesn't work. WHen running `npx chokidar`, volta said that it could not find the binary to use 🤔 I can solve this by using `yarn chokidar` instead, but I wonder if this means that `npx` wouldn't work (only if you are using volta, i imagine everything is still fine with nvm)

**Comment by noahtallen** [comment]
btw, ignore the ETK build error. (I know what the problem is; working on a fix)

File: package.json (hunk 2 above)
Comment (sirreal) [reply package.json]: Part of the `yarn` executable's behavior is to get the correct version via `.yarnrc`, isn't it? I did a test by pinning `yarn@1.21` for volta. You can see that although volta will invoke it's yarn@1.21.1 binary, yarn reports its version is `1.22.10`. I think everything will work smoothly and ultimately `.yarnrc` determines the version like I'd expect.

```
# volta list
⚡️ Currently active tools:

    Node: v12.20.1 (current @ …/calypso/package.json)
    Yarn: v1.21.1 (current @ …/calypso/package.json)
    Tool binaries available:
        wp-env (default)
        neovim-node-host (default)
        stryker (default)
        tsc, tsserver (current @ …/calypso/package.json)
        yarn-deduplicate (default)

See options for more detailed reports by running `volta list --help`.
# yarn --version
1.22.10
```

**Comment by sirreal** [comment]
> It is great that we are looking for alternatives to NVM if NVM is causing problems. Could you expand on the problems we are trying to solve with `volta`?

Volta seems to be a better tool for the job to me. When configured for a project, it does exactly what I'd want and stays out of the way otherwise. With nvm I have to think about the globals I have installed and where, and remember what node version is my default and which projects I have to run `nvm use` for (for every new terminal). It's not that nvm isn't working, it just seems inferior. I've been using volta myself (without project support!) and I've been impressed. I wrote a bit about what I like [here](https://sirre.al/2021/02/12/volta-vs-nvm-for-managing-javascript-tooling/).

> In any case, I think that we should only have one tool to do one thing. In other words, if we go ahead with this change, it should be a migration from `nvm` to `volta` with the aim of removing `nvm` from our system. Otherwise we are increasing the maintenance burden and the risk of `nvm

`nvm` is the de facto tool, but I don't think there's been any official decisions that we only support nvm. When we've upgraded node, we'd post with instructions for `nvm` and `n`. (Incidentally, node upgrade announcements don't need to happen with volta). I don't think a hard switch/mandate will go over well, but if we give folks the option and folks find it compelling, we may gain a new de facto or enough support and confidence to tell folks this is the tool they should be using. We may see other related projects follow suit based on our experience. This is the approach I'm taking here: pbABiP-hR-p2.



> One thing I noticed when testing this is that `yarn dev --sync` with ETK doesn't work. WHen running `npx chokidar`, volta said that it could not find the binary to use 🤔 I can solve this by using `yarn chokidar` instead, but I wonder if this means that `npx` wouldn't work (only if you are using volta, i imagine everything is still fine with nvm)

I've just pinned `npm` which I believe will fix this, @noahtallen will you give it a try?

File: package.json (hunk 2 above)
Comment (scinos) [reply package.json]: Seems weird and confusing to have a yarn version defined in two places. Do we really need to tell `volta` about `yarn` version?

What's the dev experience if we drop the version in `.yarnrc` and we want to update `yarn`?

**Comment by scinos** [comment]
(I'm assuming that the problem we want to solve is the need for a manual `nvm use`. I know volta does more things, but they either don't apply to us, or they haven't been mentioned as a problem yet)

> Volta seems to be a better tool for the job to me.

Then we should switch to volta, not having it just as an alternative to nvm. Honestly, having to change (and maintain) 194 files to provide a second option seems a bit too much. What would happen if we only change the root file? Also, what would happen if someone creates a new package and forgets to add that volta line? Or if they move folders and volta `extends` from a file that doesn't exists?

> I've been using volta myself (without project support!)

How does that work? Can we give folks that alternative if they don't like `nvm`?

> With nvm I have to think about the globals I have installed and where, and remember what node version is my default and which projects I have to run nvm use for (for every new terminal)

nvm has solutions for that as well, but they are not enabled out of the box: https://github.com/nvm-sh/nvm/blob/master/README.md#deeper-shell-integration. From the dev experience and causing friction point of view, I don't see a huge difference between "install volta to get rid of this problem" and "install this shell script to get rid of this problem"

> nvm is the de facto tool, but I don't think there's been any official decisions that we only support nvm.

Is the official tool in the sense that is the only tool we use for our delivery process, all our CI config is based on `nvm` and the source of truth for the Node version is encoded in `.nvmrc`. The post with the updates mentions `n` only because it was already there in the template I use to copy from (not a very good reason, I know).

I like giving devs the capacity to try out new tools and provide multiple options so they can choose the best one for their workflows. But I also believe we should keep one single source of truth (and not have node version encoded in two independant files). Looking at volta's issues looks like they won't support `.nvmrc`

--- 

To be clear: I don't like nvm. Is causing lots of duplication (check how many times we have to call `nvm.sh` in https://github.com/Automattic/wp-calypso/blob/trunk/.teamcity/settings.kts) and it is not particularly fast. I'm eager to find an alternative, and volta seems like it could be it.

But there is something I dislike more than nvm, and that's unnecessary complexity. In my opinion, having two tools for the same thing falls into the second category. That's why I think this PR should be reframed as a replacement for nvm.

File: package.json (hunk 2 above)
Comment (sirreal) [reply package.json]: If we go all-in with volta, I believe we could safely drop either one.

> What's the dev experience if we drop the version in .yarnrc and we want to update yarn?

I don't understand the question. Updating yarn via volta means running `volta pin yarn@latest` in the repo root.

Comment (scinos) [reply package.json]: Let's say we drop `.yarnrc` and we pin yarn using `volta` to a specific version (`1.22.10`).

How do update the version? I guess it is only a change in `package.json`, right?

What happens when a dev checks out a branch that points to a version they don't have locally? Does volta automatically downloads the new version?

**Comment by sirreal** [comment]
> Then we should switch to volta, not having it just as an alternative to nvm.

I agree, but I'm reluctant to make that switch based solely my experience. I'd suggest a trial period where a few folks opt-in to volta (and this PR is landed) then make a decision based on a slightly larger sample size. Team Calypso might be a good group to participate in the experiment and make the call.

> Honestly, having to change (and maintain) 194 files to provide a second option seems a bit too much.

Two clarifications about the number of changed files:
- To properly support volta, we should absolutely pin it at the root.
- We only _need_ to add a volta in directories under which folks will run node, yarn, etc. The minimum is probably the the root, client,  apps/*, and maybe packages/*.
- What I've done here is pedantically correct. Calypso is a bit unusual in that it has package.json files all over the project that don't correspond to projects/workspaces in the traditional sense. I've done these as well even though we don't expect folks to move into these directories and run yarn commands.

Only the root declaration needs to be maintained going forward. All of the other modifications to extend the root volta config need to happen exactly once (in this PR). They won't be touched again.

> > I've been using volta myself (without project support!)
> How does that work? Can we give folks that alternative if they don't like nvm?

I either apply a stashed change with the pinned versions or run `volta install node@$( cat .nvmrc )` to set the project node version as my global default version. This winds up being similar to nvm.

> nvm has solutions for that as well, but they are not enabled out of the box … I don't see a huge difference between "install volta to get rid of this problem" and "install this shell script to get rid of this problem"

The volta experience is "use volta."
The nvm experience is "use nvm and set up this shell script according to their documentation." I've never set up the additional script to auto-change versions and I don't believe it's recommended anywhere in the calypso docs.

The other big difference is that volta has optimized for speed. It seeks to minimize directory traversal and file reads as much as possible. It will also hit the network exactly one time for a given version. nvm needs to periodically check the network to make sure that it has the latest version because of its fuzzy "version-ish" specifiers. (With a fully-qualified version in our .nvmrc it may not hit the network).

> Looking at volta's issues looks like they won't support .nvmrc

Right, or more that they can't because nvm semantics are incompatible with the volta project goals.

---

> To be clear: I don't like nvm. Is causing lots of duplication (check how many times we have to call nvm.sh in https://github.com/Automattic/wp-calypso/blob/trunk/.teamcity/settings.kts) and it is not particularly fast. I'm eager to find an alternative, and volta seems like it could be it.

We may be able to replace volta with nvm for better results. The volta executable can be installed in the environment and/or cached. The installed versions can be cached, and the setup should work without calling nvm over and over.

> In my opinion, having two tools for the same thing falls into the second category. That's why I think this PR should be reframed as a replacement for nvm.

I'd be fine with that, except I don't want to make a decision on what this project should be using especially since I'm not contributing as frequently to Calypso. I don't think I'm a good authority to make this decision (alone) and I certainly don't want to be the sole party responsible for helping folks to adjust.

File: package.json (hunk 2 above)
Comment (sirreal) [reply package.json]: > How do update the version? I guess it is only a change in package.json, right?

Yep.

> What happens when a dev checks out a branch that points to a version they don't have locally? Does volta automatically downloads the new version?

Yes.

**Comment by creativecoder** [comment]
I believe moving more projects to volta would also solve problems with initializing multiple projects in a8c-wp-env (e.g. 80-gh-Automattic/a8c-wp-env)

**Comment by sirreal** [comment]
A quick update:

- We've had really good results using Volta on Redpop. No complaints.
- The one difficulty is that Node wasn't publishing native builds for new Apple chips (M1). nvm would compile node in these cases, whereas volta would download the non-native binary and it would run through rosetta which was bad for performance. With Node 16, native binaries are published and the latest volta version will use them correctly. [Node plans to add native binaries to Node 14 (current LTS) but hasn't done so yet](https://github.com/nodejs/node/issues/37309).

I continue to believe volta provides a better experience and the only concern I have is resolved by either a Node release or an upgrade of the Calypso node version. Upgrading Calypso to Node 16 should be fine, it's the current Node release and will become current LTS in October.

**Comment by sirreal** [comment]
Rebased and added extends to all the new package.json I found.

**Review by creativecoder — APPROVED** [review]
Looks like this needs more conflicts resolved before merging but wanted to add my ✅.

This tests as a much nicer way of managing node versions, and solves some of the struggles I have with nvm and node versions when running node commands as tasks in VS Code.

**Comment by creativecoder** [comment]
FYI, I'm proposing the same change in Jetpack: https://github.com/Automattic/jetpack/pull/19705

**Comment by noahtallen** [comment]
I do like the ergonomics of Volta a lot, and have been using it exclusively recently, even without these changes. I wonder what our next steps should be? Do we need to commit to nvm or volta, or can we support both at the same time? (I don't see supporting volta as having a high cost, since it just involves bumping the version in the root package.json whenever we update node.)

Relatedly... I _really_, _really_ wish they had not gone with this approach for monorepos. It seems to be the largest source of concern for project adoption since it increases the surface area by so much.

**Comment by scinos** [comment]
>  I wonder what our next steps should be? Do we need to commit to nvm or volta, or can we support both at the same time?

I still think we should draft a plan to do a full transition from NVM to Volta. IMO, at the very least we need (in no particular order):

- Change our Docker images to use Volta instead of NVM
- Write P2s explaining the difference, defining the transition plan and in general educating our devs about the new tool.
- Land this PR.

Nice to have:
- Clean up all those package.json. As @sirreal said, only a few of those will actually need volta (let's say only those that define 'scripts'?). Eventually get rid of the "fake" package.json files, this is not the first time they cause troubles.
- Detect when someone is using node from nvm, and warn them (example: have a pre-install hook that warns if the node binary is from `~/.nvm/versions/node/v*/bin/node`)

**Comment by creativecoder** [comment]
> I still think we should draft a plan to do a full transition from NVM to Volta. IMO, at the very least we need (in no particular order): ...

That all seems good to me.

Note that I still have nvm installed and volta seems to take precedence when running shell commands for which version of node and any npm installed binaries are used.

**Comment by sirreal** [comment]
I can draft or help with a p2 post. I'd recommend upgrading to Node 16 first (p7H4VZ-31V-p2) so everyone gets [native Node binaries with Volta](https://github.com/volta-cli/volta/pull/974).

**Comment by sirreal** [comment]
This is very stale, I'll close it now. If there's interest and a desire to move forward with this I'd be happy to get this ready again.

**Comment by noahtallen** [comment]
Totally. I really wish they would just support reading from a higher-level .json file! I also looked into manual pinning -- like if I could achieve automatic version switching via a file I could add to .gitignore, but that also isn't supported :(

So I'm back to nvm for now, just because both Gutenberg and Calypso use it.

**Comment by scinos** [comment]
> I really wish they would just support reading from a higher-level .json file

For the record, they kind of do, with `extends`. [Docs](https://docs.volta.sh/advanced/workspaces).

**Comment by noahtallen** [comment]
Oh hey! Hope things are going great :D That's what this PR was exploring originally

**Comment by creativecoder** [comment]
I'm following this issue in the volta repo, but there hasn't been much movement: https://github.com/volta-cli/volta/issues/282

**Pull request closed without merging**